Flutter Link Mode (#1693)
WalletConnect/walletconnect-docs: docs/walletkit/flutter/link-mode.mdx

import PlatformTabs from '../../components/PlatformTabs'
import PlatformTabItem from '../../components/PlatformTabItem'

# Link Mode

> **WalletKit Link Mode** is a low latency mechanism for transporting 1-click auth requests and session requests over universal links, eliminating the need for a WebSocket connection with the Relay. This significantly enhances the user experience when connecting native dApps to native wallets by reducing the latency associated with networking connections, especially when the user has an unstable internet connection.

Reviewer: ```suggestion
**WalletKit Link Mode** is a low latency mechanism for transporting One Click Auth requests and session requests over universal links, eliminating the need for a WebSocket connection with the Relay. This significantly enhances the user experience when connecting native dApps to native wallets by reducing the latency associated with networking connections, especially when the user has an unstable internet connection.
```

> By enabling it, wallet and dapp will communicate through declared Universal Links on iOS and/or App Links on Android.

Reviewer: isn't it Universal Link for Android as well?

Author: Nope, Android calls them App Links https://developer.android.com/training/app-links

import PlatformTabs from '../../components/PlatformTabs'
import PlatformTabItem from '../../components/PlatformTabItem'

# Link Mode

> **WalletKit Link Mode** is a low latency mechanism for transporting One-Click Auth requests over Universal Links, eliminating the need for a WebSocket connection with the Relay. This significantly enhances the user experience when connecting native dApps to native wallets by reducing the latency associated with networking connections, especially when the user has an unstable internet connection.

Reviewer 2: ```suggestion
**WalletKit Link Mode** is a low latency mechanism for transporting One-Click Auth requests over Universal Links, eliminating the need for a WebSocket connection with the Relay. This significantly enhances the user experience when connecting native dApps to native wallets by reducing the latency associated with network connections, especially when the user has an unstable internet connection.
```

Grammar fix

> By enabling it, wallet and dapp will communicate through declared Universal Links on iOS and/or App Links on Android **even without internet connection.**

Reviewer 2: ```suggestion
By enabling it, the wallet and dapp will communicate through declared Universal Links on iOS and/or App Links on Android **even without an internet connection.**
```

Grammar fix

:::caution
Link Mode is currently in **Beta** for Flutter. You should install [v2.4.0-beta01](https://pub.dev/packages/walletconnect_flutter_v2/versions/2.4.0-beta01) to try it out
:::

### How to enable it:

> 1. Add an Universal Link for your wallet in the **Explorer** tab of your [**Cloud project configuration**](https://cloud.walletconnect.com/sign-in), under **Mobile Linking** section

Reviewer 2: ```suggestion
1. Add an Universal Link for your wallet in the **Explorer** tab of your [**Cloud project configuration**](https://cloud.walletconnect.com/sign-in), under the **Mobile Linking** section
```

Grammar fix

2. Configure your `PairingMetadata`'s `redirect:` object with that Universal Link

3. Set the `linkMode` property to `true`:

```dart
final web3Wallet = Web3Wallet(
  projectId: '123.....',
  metadata: const PairingMetadata(
    ...
    redirect: Redirect(
      native: 'yourwalletscheme://',
      universal: 'https://your_wallet_universal_link.com/open',
      linkMode: true,
    ),
  ),
);
```

> Once everything is properly configured and the user interacts with a Link Mode supporting dApp, your Wallet will receive requests over it.

Reviewer 2: ```suggestion
Once everything is properly configured, and the user interacts with a Link Mode-supporting dApp, your wallet will receive requests through it.
```

Minor fix for better readability.

> On Flutter there are different plugins that would help you integrate Universal/App Links but, no matter which one you choose, it is important that, when capturing an incoming link, you pass it to WalletKit so it can process it:

Reviewer 2: ```suggestion
In Flutter, there are various plugins that can help you integrate Universal/App Links. However, regardless of which one you choose, it is crucial that, when capturing an incoming link, you pass it to WalletKit so it can process the request.
```

Improving readability

```dart
void _onLinkCaptured(String link) async {
  await web3Wallet.dispatchEnvelope(link);
}
```

### Platform specifics:

<PlatformTabs
	groupId="w3m_flutter"
	activeOptions={["ios","android"]}
>
<PlatformTabItem value="ios">

> 1. Ensure to handle incoming universal links in different methods of `AppDelegate` or `SceneDelegate`. 

Reviewer 2: ```suggestion
1. Ensure that you handle incoming Universal Links in the appropriate methods of `AppDelegate` or `SceneDelegate`.
```

Grammar fix

> 2. Ensure that you enabled the Associated Domains Capability in your XCode project and that your Universal Link is properly configured. _(Depending on the preivous states of your Provisioning Profiles it may be required to update/create new ones)_

Reviewer 2: ```suggestion
2. Ensure that you have enabled the Associated Domains Capability in your XCode project and that your Universal Link is properly configured. _(Depending on the previous states of your Provisioning Profiles it may be necessary to update or create new ones)_
```

Grammar fix

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>com.apple.developer.associated-domains</key>
  <array>
    <string>applinks:your_wallet_universal_link.com</string>
  </array>
</dict>
</plist>
```
3. Update/Create your domain's `.well-known/apple-app-site-association` file accordingly.

For more information on how to configure universal links for your app, refer to the [Apple Documentation](https://developer.apple.com/documentation/xcode/allowing-apps-and-websites-to-link-to-your-content?language=swift).<br />
For a debugging guide, visit the [Debugging Universal Links](https://developer.apple.com/documentation/technotes/tn3155-debugging-universal-links) page.<br />

You can check our Flutter Sample Wallet's [AppDelegate file](https://github.com/WalletConnect/WalletConnectFlutterV2/blob/develop/example/wallet/ios/Runner/AppDelegate.swift) as a reference.

</PlatformTabItem>
<PlatformTabItem value="android">

> 1. Ensure to handle incoming App Links in your Activity's `onCreate` method and in `onNewIntent` callback.

Reviewer 2: ```suggestion
1. Ensure that you handle incoming App Links in your Activity's `onCreate` method and in `onNewIntent` callback.
```

Grammar fix

2. Ensure that your App Link is properly configured in your app's `AndroidManifest.xml` file with the `autoVerify` set to `true`:
```xml
<intent-filter android:autoVerify="true">
  <action android:name="android.intent.action.VIEW" />

  <category android:name="android.intent.category.DEFAULT" />
  <category android:name="android.intent.category.BROWSABLE" />

  <data android:scheme="https" />
  <data android:host="your_wallet_universal_link.com" />
  <data android:pathPattern="/open" />
</intent-filter>
```
3. Update/Create your domains's `.well-known/assetlinks.json` file accordingly

For more information on how to configure app links for your app, refer to the [Android Documentation](https://developer.android.com/training/app-links/verify-android-applinks).<br />
For enabling links to app content check [this](https://developer.android.com/training/app-links/deep-linking) documentation page.<br />
For more information on how to interact with other apps using intents, see [Android Intent Documentation](https://developer.android.com/training/basics/intents).

> You can check our Flutter Sample Wallet's [MainActivity file](https://github.com/WalletConnect/WalletConnectFlutterV2/blob/develop/example/wallet/android/app/src/main/kotlin/com/example/wallet/MainActivity.kt) as a reference.

Reviewer 2: ```suggestion
You can check out tge Flutter Sample Wallet's [MainActivity file](https://github.com/WalletConnect/WalletConnectFlutterV2/blob/develop/example/wallet/android/app/src/main/kotlin/com/example/wallet/MainActivity.kt) as a reference.
```

Grammar fix
